flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -1305,6 +1308,64 @@ public void testAllocationUpdatesIgnoredIfSlotMarkedAsAllocatedAfterSlotReport()
+                new CompletableFuture<>();
+        final CompletableFuture<Acknowledge> firstSlotRequestAcknowledgeFuture =
+                new CompletableFuture<>();
+        final Iterator<CompletableFuture<Acknowledge>> slotRequestAcknowledgeFutures =
+                Arrays.asList(
+                                firstSlotRequestAcknowledgeFuture,
+                                new CompletableFuture<Acknowledge>())
+                        .iterator();
+
+        final TestingTaskExecutorGateway taskExecutorGateway =
+                new TestingTaskExecutorGatewayBuilder()
+                        .setRequestSlotFunction(
+                                requestSlotParameters -> {
+                                    firstSlotAllocationIdFuture.complete(requestSlotParameters.f2);
+                                    return slotRequestAcknowledgeFutures.next();
+                                })
+                        .createTestingTaskExecutorGateway();
+
+        try (final DeclarativeSlotManager slotManager =
+                createDeclarativeSlotManagerBuilder()
+                        .buildAndStart(
+                                ResourceManagerId.generate(),
+                                ComponentMainThreadExecutorServiceAdapter.forMainThread(),
+                                new TestingResourceActionsBuilder().build())) {
+
+            final TaskExecutorConnection taskExecutionConnection =
+                    createTaskExecutorConnection(taskExecutorGateway);
+            final SlotReport slotReport =
+                    createSlotReport(taskExecutionConnection.getResourceID(), 1);
+            final SlotID slotId = Iterators.getOnlyElement(slotReport.iterator()).getSlotID();
+
+            // register task executor
+            slotManager.registerTaskManager(
+                    taskExecutionConnection, slotReport, ResourceProfile.ANY, ResourceProfile.ANY);
+            slotManager.reportSlotStatus(
+                    taskExecutionConnection.getInstanceID(),
+                    createSlotReport(taskExecutionConnection.getResourceID(), 1));
+
+            // triggers the allocation of a slot
+            final JobID firstJobId = new JobID();
+            slotManager.processResourceRequirements(createResourceRequirements(firstJobId, 1));
+            // clear requirements immediately to ensure the slot will not get re-allocated to the
+            // same job
+            slotManager.processResourceRequirements(
+                    ResourceRequirements.empty(firstJobId, "foobar"));
+            // when the slot is freed it will be re-assigned to this second job
+            slotManager.processResourceRequirements(createResourceRequirements(new JobID(), 1));
+
+            slotManager.freeSlot(slotId, firstSlotAllocationIdFuture.get());
+
+            // acknowledge the first allocation
+            firstSlotRequestAcknowledgeFuture.complete(Acknowledge.get());
```

Comment:
Is this test missing an assertion?

Reply:
No (the call would fail with an exception), but it lacks clarity which was also raised in the PR review #15288.